src/main/java/com/example/GymProject/service/TraineeService.java

```diff
@@ -1,56 +1,164 @@
-        trainee.setLastName(newTrainee.getLastName());
-        trainee.setAddress(newTrainee.getAddress());
-        trainee.setDob(newTrainee.getDob());
-        trainee.setActive(newTrainee.getActive());
+    public void activateDeactivateTrainee(String username, boolean isActive, String password) {
+        Assert.notNull(username, "Username cannot be null");
+        authenticate(username, password);
+        Trainee trainee = traineeDao.getTraineeByUsername(username);
+        if (trainee != null) {
+            logger.info("Activation/Deactivation was successfully performed for trainee {}", username);
+            trainee.getUser().setIsActive(isActive);
+            userService.updateUser(entityMapper.toUserDto(trainee.getUser()));
+        }
+    }
+
+    public List<TrainingDto> getTraineeTrainings(String username, Date fromDate, Date toDate, String trainerName, String trainingType, String password) {
+        Assert.notNull(username, "Username cannot be null");
+        authenticate(username, password);
+        List<Training> trainings = traineeDao.getTraineeTrainings(username, fromDate, toDate, trainerName, trainingType);
+        logger.info("Getting trainings for trainee {}", username);
+        return trainings.stream()
+                .map(entityMapper::toTrainingDto)
+                .collect(Collectors.toList());
+    }
 
-        memoryStorage.getTraineeDao().update(key, trainee);
+    public void updateTraineeTrainers(String username, Set<TrainerDto> trainerDtos, String password) {
+        Assert.notNull(username, "Username cannot be null");
+        authenticate(username, password);
+        Trainee trainee = traineeDao.getTraineeByUsername(username);
+        logger.info("Updating trainers for trainee {}", username);
+        Set<Trainer> trainers = trainerDtos.stream()
+                .map(entityMapper::toTrainer)
+                .collect(Collectors.toSet());
+        trainee.setTrainers(trainers);
+        traineeDao.updateTrainee(trainee);
     }
 
-    public void deleteTrainee(String key) {
-        memoryStorage.getTraineeDao().delete(key);
+    @Transactional(readOnly = true)
+    public List<TrainerDto> getUnassignedTrainers(String traineeUsername, String password) {
+        authenticate(traineeUsername, password);
+        Trainee trainee = traineeDao.getTraineeByUsername(traineeUsername);
+        List<Trainer> allTrainers = trainerDao.getAllTrainers();
+
+        List<Trainer> assignedTrainers = trainee.getTrainers().stream()
+                .collect(Collectors.toList());
+
+        List<Trainer> unassignedTrainers = allTrainers.stream()
+                .filter(trainer -> !assignedTrainers.contains(trainer))
+                .collect(Collectors.toList());
+
+        return unassignedTrainers.stream()
+                .map(entityMapper::toTrainerDto)
+                .collect(Collectors.toList());
+    }
+
+    public void authenticate(String username, String password) {
```

Comment:
I guess better to refactor this to boolean isAuthenticated... and return true if match, false otherwise. Now you throw some Exception which could be not catched properly in your services. Or am I wrong?

Reply:
Yes, your'e right, I will change this method to return boolean.